main.tf

```diff
@@ -84,6 +86,59 @@ resource "aws_ecs_service" "consul" {
   ]
 }
 
+resource "aws_ecs_service" "consul_primary" {
+  count           = "${length(var.ecs_cluster_ids)  > 1 ? 1 : 0}"
+  name            = "consul-${var.env}-primary"
+  cluster         = "${var.ecs_cluster_ids[0]}"
+  task_definition = "${aws_ecs_task_definition.consul.arn}"
+  desired_count   = "${var.cluster_size * 2 }"                    # This is not awesome, it lets new AS groups get added to the cluster before destruction.
```

Comment:
ok, i think i just have a mental block about this.  why do we want `desired_count` to be 2x `cluster_size`?  maybe let's go to the whiteboard and draw a picture?

Reply:
We can totally hit the WB!!   I'll put the short answer here though. First, here's the TF plan when a new AMI drops:

```
-/+ module.ecs_consul_secondary.aws_autoscaling_group.ecs (new resource required)
      id:                                                "asg-ecs-consul-dev-secondary-00e55c71e17ce2378895402f1c-00e55c71e17ce2378895402f1d" => <computed> (forces new resource)
      arn:                                               "arn:aws:autoscaling:us-east-1:332913666507:autoScalingGroup:46acaf89-bd84-46ca-a406-6a7a0ac94d69:autoScalingGroupName/asg-ecs-consul-dev-secondary-00e55c71e17ce2378895402f1c-00e55c71e17ce2378895402f1d" => <computed>
      default_cooldown:                                  "300" => <computed>
      desired_capacity:                                  "3" => "3"
      force_delete:                                      "false" => "false"
      health_check_grace_period:                         "300" => "300"
      health_check_type:                                 "EC2" => <computed>
      launch_configuration:                              "ecs-consul-dev-secondary-00e55c71e17ce2378895402f1c" => "${aws_launch_configuration.ecs.name}"
      load_balancers.#:                                  "0" => <computed>
      max_size:                                          "10" => "10"
      metrics_granularity:                               "1Minute" => "1Minute"
      min_size:                                          "1" => "1"
      name:                                              "asg-ecs-consul-dev-secondary-00e55c71e17ce2378895402f1c-00e55c71e17ce2378895402f1d" => <computed>
      name_prefix:                                       "asg-ecs-consul-dev-secondary-00e55c71e17ce2378895402f1c-" => "asg-${aws_launch_configuration.ecs.name}-" (forces new resource)
      protect_from_scale_in:                             "false" => "false"
      tags.#:                                            "3" => "3"
      tags.0.%:                                          "3" => "3"
      tags.0.key:                                        "Name" => "Name"
      tags.0.propagate_at_launch:                        "1" => "1"
      tags.0.value:                                      "consul-dev-secondary ECS Node" => "consul-dev-secondary ECS Node"
      tags.1.%:                                          "3" => "3"
      tags.1.key:                                        "consul_server" => "consul_server"
      tags.1.propagate_at_launch:                        "1" => "1"
      tags.1.value:                                      "true" => "true"
      tags.2.%:                                          "3" => "3"
      tags.2.key:                                        "dev_consul_server" => "dev_consul_server"
      tags.2.propagate_at_launch:                        "1" => "1"
      tags.2.value:                                      "true" => "true"
      target_group_arns.#:                               "0" => <computed>
      termination_policies.#:                            "3" => "3"
      termination_policies.0:                            "OldestLaunchConfiguration" => "OldestLaunchConfiguration"
      termination_policies.1:                            "ClosestToNextInstanceHour" => "ClosestToNextInstanceHour"
      termination_policies.2:                            "Default" => "Default"
      vpc_zone_identifier.#:                             "3" => "3"
      vpc_zone_identifier.3244149418:                    "subnet-a0464c8d" => "subnet-a0464c8d"
      vpc_zone_identifier.4196667900:                    "subnet-3927dd71" => "subnet-3927dd71"
      vpc_zone_identifier.709250968:                     "subnet-5c444507" => "subnet-5c444507"
      wait_for_capacity_timeout:                         "10m" => "10m"

  ~ module.ecs_consul_secondary.aws_autoscaling_lifecycle_hook.graceful_shutdown_asg_hook
      autoscaling_group_name:                            "asg-ecs-consul-dev-secondary-00e55c71e17ce2378895402f1c-00e55c71e17ce2378895402f1d" => "${aws_autoscaling_group.ecs.name}"

-/+ module.ecs_consul_secondary.aws_launch_configuration.ecs (new resource required)
      id:                                                "ecs-consul-dev-secondary-00e55c71e17ce2378895402f1c" => <computed> (forces new resource)
      associate_public_ip_address:                       "false" => "false"
      ebs_block_device.#:                                "1" => "1"
      ebs_block_device.1563169761.delete_on_termination: "true" => "true"
      ebs_block_device.1563169761.device_name:           "/dev/xvdcz" => "/dev/xvdcz"
      ebs_block_device.1563169761.encrypted:             "false" => <computed>
      ebs_block_device.1563169761.iops:                  "0" => <computed>
      ebs_block_device.1563169761.snapshot_id:           "" => <computed>
      ebs_block_device.1563169761.volume_size:           "22" => "22"
      ebs_block_device.1563169761.volume_type:           "gp2" => "gp2"
      ebs_optimized:                                     "false" => <computed>
      enable_monitoring:                                 "true" => "true"
      iam_instance_profile:                              "tf-ECSProfile-consul-dev-secondary-000926b8bd9a727d337c561251" => "tf-ECSProfile-consul-dev-secondary-000926b8bd9a727d337c561251"
      image_id:                                          "ami-9eb4b1e5" => "ami-ec33cc96" (forces new resource)
      instance_type:                                     "t2.micro" => "t2.micro"
      key_name:                                          "rk-utility" => "rk-utility"
      name:                                              "ecs-consul-dev-secondary-00e55c71e17ce2378895402f1c" => <computed>
      name_prefix:                                       "ecs-consul-dev-secondary-" => "ecs-consul-dev-secondary-"
      root_block_device.#:                               "0" => <computed>
      security_groups.#:                                 "1" => "1"
      security_groups.521627524:                         "sg-40109132" => "sg-40109132"
      user_data:                                         "5081c5ba5e1fb50ed9218bbe84161eb52f995eb6" => "5081c5ba5e1fb50ed9218bbe84161eb52f995eb6"


Plan: 2 to add, 1 to change, 2 to destroy.
```

So when Terraform creates a new ASG and new launch config, there needs to be a high enough number of tasks to grow the cluster - i.e start new docker images on the new instances, to allow consul to  migrate to the new instances *before* the old ASG is destroyed. Otherwise, ECS says "ok, I've now got 10 instances, and need 5 running tasks which i *have* so everything is cool"  *until* the old ASG is removed, at which point ECS will engage and spin up new tasks, but at this point, Consul has lost 50% of it's cluster and can't recover.  However, if the task number is high, when the new instance spin up, it *immediately* grows the cluster into them, stabilizes and when the old ASG is removed, had enough nodes to continue working. 

Ha, I know I said that was going to be a short answer... 🤕

Reply:
ok, NOW i've got it.  thx.  :shipit: and we'll put the new locking rolling deploy strategy in the background.